src/gazebo_hrim_joint_state_publisher.cpp

```diff
@@ -0,0 +1,187 @@
+
+  /// Pointer to the update event connection.
+  gazebo::event::ConnectionPtr update_connection_;
+};
+
+GazeboRosJointStatePublisher::GazeboRosJointStatePublisher()
+: impl_(std::make_unique<GazeboRosJointStatePublisherPrivate>())
+{
+}
+
+GazeboRosJointStatePublisher::~GazeboRosJointStatePublisher()
+{
+}
+
+void GazeboRosJointStatePublisher::Load(gazebo::physics::ModelPtr model, sdf::ElementPtr sdf)
+{
+  // ROS node
+  impl_->ros_node_ = gazebo_ros::Node::Get(sdf);
+
+  // Joints
+  if (!sdf->HasElement("joint_name")) {
+    RCLCPP_ERROR(impl_->ros_node_->get_logger(), "Plugin missing <joint_name>s");
+    impl_->ros_node_.reset();
+    return;
+  }
+
+  sdf::ElementPtr joint_elem = sdf->GetElement("joint_name");
+  while (joint_elem) {
+    auto joint_name = joint_elem->Get<std::string>();
+
+    auto joint = model->GetJoint(joint_name);
+    if (!joint) {
+      RCLCPP_ERROR(impl_->ros_node_->get_logger(), "Joint %s does not exist!", joint_name.c_str());
+    } else {
+      impl_->joints_.push_back(joint);
+      RCLCPP_INFO(impl_->ros_node_->get_logger(), "Going to publish joint [%s]",
+        joint_name.c_str() );
+    }
+
+    joint_elem = joint_elem->GetNextElement("joint_name");
+  }
+
+  if (impl_->joints_.empty()) {
+    RCLCPP_ERROR(impl_->ros_node_->get_logger(), "No joints found.");
+    impl_->ros_node_.reset();
+    return;
+  }
+
+  // Update rate
+  double update_rate = 100.0;
+  if (!sdf->HasElement("update_rate")) {
+    RCLCPP_INFO(impl_->ros_node_->get_logger(), "Missing <update_rate>, defaults to %f",
+      update_rate);
+  } else {
+    update_rate = sdf->GetElement("update_rate")->Get<double>();
+  }
+
+  if (update_rate > 0.0) {
+    impl_->update_period_ = 1.0 / update_rate;
+  } else {
+    impl_->update_period_ = 0.0;
+  }
+
+  impl_->last_update_time_ = model->GetWorld()->SimTime();
+
+  // Joint state publisher
+  impl_->joint_state_pub_ = impl_->ros_node_->create_publisher<hrim_composite_arm_msgs::msg::JointStates>(
+    "joint_states", 1000);
+
+  // Callback on every iteration
+  impl_->update_connection_ = gazebo::event::Events::ConnectWorldUpdateBegin(
+    std::bind(&GazeboRosJointStatePublisherPrivate::OnUpdate, impl_.get(), std::placeholders::_1));
+}
+
+void GazeboRosJointStatePublisherPrivate::OnUpdate(const gazebo::common::UpdateInfo & info)
+{
+  gazebo::common::Time current_time = info.simTime;
+
+  // If the world is reset, for example
+  if (current_time < last_update_time_) {
+    RCLCPP_INFO(ros_node_->get_logger(), "Negative sim time difference detected.");
+    last_update_time_ = current_time;
+  }
+
+  // Check period
+  double seconds_since_last_update = (current_time - last_update_time_).Double();
+
+  if (seconds_since_last_update < update_period_) {
+    return;
+  }
+
+  // Populate message
+  hrim_composite_arm_msgs::msg::JointStates joint_state;
+
+  joint_state.header.stamp = gazebo_ros::Convert<builtin_interfaces::msg::Time>(current_time);
```

Comment:
Wrong!

Reply:
My bad, on it.

Reply:
Gazebo does his own conversion of the simulation's time to the ROS 2 one. I'd say this didn't throw exceptions as the definitions of the Time structure in ROS 2 and HRIM are comparable.

This, however, is kinda cheating, so for now a manual conversion of this data is the quickest way to go.

I propose revisiting this at a later date to generate the necessary gazebo packages making use of HRIM messages natively. I'd say this is overshooting quite a bit for now and we should wait until everything is more fleshed out, as this would probably entail rewriting a huge amount of code in gazebo.